README.md
# ideal-cli

Frontend cli

## 一、Features

* Support build（Webpack）
  * development
  * production
  * library
* code format
* pre hooks check
* standard version
* dll
* build analysis
* browser compatibility
* commit lint

### Coming soon

- [ ] Unit test（Jest）

## 二、Usages

```bash
npm i ideal-cli@latest -D -E

# First init configuration
npx ideal-cli init -h
```

## 三、Configurations

### 1. package.json

```json
{
  "scripts": {
    "start": "ideal-cli start",
    "build": "ideal-cli build",
    "build-lib": "ideal-cli build-lib"
  }
}
```

### 2. browserslist

Add configuration to ```package.json```

```json
{
  "browserslist": [
    "> 1%",
    "not dead",
    "not op_mini all",
    "Firefox ESR",
    "Chrome >= 68",
    "IE 11",
    "Firefox >= 56",
    "Opera >= 48",
    "Safari >= 11"
  ]
}
```

### 3. babelrc

If project uses __react__, you should config local ```.babelrc```:

```json
{
  "presets": [
    "react-app"
  ]
}
```

### 4. lint

```bash
npx eslint --init
```

```json
{
  "lint": "eslint 'src/**/*.[tj]s?(x)'",
  "lint-fix": "npm run lint -- --fix",
  "format": "prettier 'src/**/*.[tj]s?(x)' --check --write && npm run lint-fix"
}
```

### 5. version

> major.minor.patch

```bash
npm run release -- --first-release
npm run release -- --release-as (major|minor|patch)
```

[More Info](https://www.npmjs.com/package/standard-version)

### 6. Dll

Config __package.json__

```json
{
  "scripts": {
    "prestart": "rimraf dist && ideal-cli dll",
    "prebuild": "rimraf dist && ideal-cli dll"
  }
}
```

Config your __index.html__.

```html
<body>
  <!-- contents -->

  <!-- Inject DLL -->
  <script src="%INJECT_DLL%"></script>
</body>
```

Config your __project.config.js__.

```js
{
  dllVendors: ['react', 'react-dom']
}
```

## 四、Optimization

* [code splitting](https://webpack.docschina.org/guides/code-splitting/)
* [lazy loading](https://webpack.docschina.org/guides/lazy-loading/)
* [tree shaking](https://webpack.docschina.org/guides/tree-shaking/)
* [Shimming](https://webpack.docschina.org/guides/shimming/)
* Cache
* Concurrency

## 五、Bundle analysis

```bash
npm run analyze
```

Analysis Website

* <http://webpack.github.io/analyse/>
* <https://webpack.jakoblind.no/optimize/>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.0.0-alpha.11](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/compare/v1.0.0-alpha.10...v1.0.0-alpha.11) (2020-10-24)


### Bug Fixes

* dev dll path ([b91ad4a](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/b91ad4adf5c8f1b05a30c3f8bde130fc751fd728))

## [1.0.0-alpha.10](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/compare/v1.0.0-alpha.9...v1.0.0-alpha.10) (2020-10-24)


### Bug Fixes

* dll path ([df1df08](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/df1df083959ae2e86759b6d8ddd80dc5307c6d18))

## [1.0.0-alpha.9](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/compare/v1.0.0-alpha.8...v1.0.0-alpha.9) (2020-10-22)


### Bug Fixes

* reback alpha.5 ([6d38180](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/6d38180eaa9ff93522d124072a37fbbeb5daa183))

## [1.0.0-alpha.8](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/compare/v1.0.0-alpha.7...v1.0.0-alpha.8) (2020-10-22)


### Bug Fixes

* import file ([8bae8e4](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/8bae8e47f6dc6f3cfd9d1e5eb92a7cb6251080c2))

## [1.0.0-alpha.7](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/compare/v1.0.0-alpha.5...v1.0.0-alpha.7) (2020-10-22)


### Features

* add rollup ([ddb644b](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/ddb644b63f6db266089ed8b874f7a936d5f30c9e))


### Bug Fixes

* dll build path error ([bc4c2ef](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/bc4c2efa8b48b051f6314468acfd27bab0390249))

## [1.0.0-alpha.6](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/compare/v1.0.0-alpha.5...v1.0.0-alpha.6) (2020-10-22)


### Features

* add rollup ([ddb644b](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/ddb644b63f6db266089ed8b874f7a936d5f30c9e))


### Bug Fixes

* dll build path error ([bc4c2ef](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/bc4c2efa8b48b051f6314468acfd27bab0390249))

## [1.0.0-alpha.5](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/compare/v1.0.0-alpha.4...v1.0.0-alpha.5) (2020-08-04)


### Features

* add commit lint ([048634e](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/048634e9af8e7f782e9c88195d286252b9b23493))

## [1.0.0-alpha.4](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/compare/v1.0.0-alpha.3...v1.0.0-alpha.4) (2020-07-26)


### Features

* optimise lib configurationn ([3da8e58](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/3da8e58158f92baa1051af3441f9e69415fed425))

## [1.0.0-alpha.3](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/compare/v1.0.0-alpha.2...v1.0.0-alpha.3) (2020-07-26)

## [1.0.0-alpha.2](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/compare/v1.0.0-alpha.1...v1.0.0-alpha.2) (2020-07-26)

## 1.0.0-alpha.1 (2020-07-25)


### Features

* add dll ([618c8c5](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/618c8c541a1666fe384bfdaf86fcbba19dc3d2cf))
* add init stage ([848da5f](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/848da5fcba4a991b0665eae480cf3bfebcab5c68))
* add lint ([230b169](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/230b1695b04a4abbcbcab8d3a95e8f8aefb2f708))
* init product ([c6c8f02](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/c6c8f02acf661516cac89178ef5626ee8a0d42a0))
* optimise code && add bundle analysis ([81c030d](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/81c030dbfc0dbcb45b746ee75f262e8c7a76e4c0))
* optimise files ([180792a](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/180792a348c6cdee23aae46432904149e4b8da6c))
* optimise init ([bf273c8](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/bf273c8cc69ea907688877251948609a24190460))
* support less ([2728a33](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/2728a330291048a548382f05dd347402cb1ffd70))
* update dev config ([65bad37](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/65bad374915f59f318fee52caccf8322c64147ce))
* update prod & lib config ([998bd3e](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/998bd3e0bace8d5bd4257203380a3bee0b850a50))
* update tsconfig ([b19b5cb](https://taylorpzreal.github.com/TaylorPzreal/ideal-cli/commit/b19b5cbc717c3ceabc26494ad4577e85cca38163))


config/index.js
const path = require('path');
const fs = require('fs');
const packagePath = path.resolve(process.cwd(), 'package.json');

if (!fs.existsSync(packagePath)) {
  console.log('Does`t get package.json from your project');
  return;
}

const packageJSON = require(packagePath);

const scriptsObject = {
  start: 'NODE_ENV=development ideal-cli start',
  build: 'NODE_ENV=production ideal-cli build',
  prestart: 'rimraf dist && NODE_ENV=development ideal-cli dll',
  prebuild: 'rimraf dist && NODE_ENV=production ideal-cli dll',
  'build-lib': 'tsc && NODE_ENV=production ideal-cli build-lib',
  prepublishOnly: 'npm run build-lib',
  lint: 'eslint \'src/**/*.[jt]s?(x)\'',
  'lint-fix': 'npm run lint -- --fix',
  format: 'prettier \'src/**/*.[tj]s?(x)\' --check --write && npm run lint-fix',
  release: 'standard-version',
  analyze: 'webpack-bundle-analyzer analysis/stats.json',
  commit: 'commit',
}

function addScript(result, script) {
  if (!result.scripts[script]) {
    result.scripts[script] = scriptsObject[script];
  } else {
    console.log(`Ignore script: ${script}`);
  }
}

function updatePackage(options = {}) {
  const { isLibrary } = options;
  const { browserslist, husky, 'lint-staged': lintStaged } = packageJSON;
  let result = Object.assign({}, packageJSON);

  // update scripts
  const scriptsKeys = Object.keys(scriptsObject);
  let applyKeys = [];
  const commonKeys = ['lint', 'lint-fix', 'format', 'release', 'analyze', 'commit'];
  if (isLibrary) {
    applyKeys = scriptsKeys.filter((script) => ['build-lib', 'prepublishOnly'].includes(script));
  } else {
    applyKeys = scriptsKeys.filter((script) => ['start', 'build', 'prestart', 'prebuild'].includes(script));
  }
  
  [...applyKeys, ...commonKeys].forEach((script) => {
    addScript(result, script);
  });

  // update browserslist
  if (!browserslist) {
    result.browserslist = [];
    result.browserslist.push(
      "> 1%",
      "not dead",
      "not op_mini all",
      "Firefox ESR",
      "Chrome >= 68",
      "IE 11",
      "Firefox >= 56",
      "Opera >= 48",
      "Safari >= 11"
    );
  } else {
    console.log('Ignore browserslist configuration');
  }

  // add hooks
  const hooks = {
    "husky": {
      "hooks": {
        "pre-commit": "lint-staged",
        "commit-msg": "commitlint -E HUSKY_GIT_PARAMS"
      }
    },
    "lint-staged": {
      "src/**/*.[jt]s?(x)": [
        "npm run format"
      ],
    }
  }

  if (!husky && !lintStaged) {
    result = Object.assign({}, result, {...hooks});
  }

  fs.writeFile(packagePath, JSON.stringify(result, null, 2), { encoding: 'utf8' }, (err) => {
    if (err) {
      console.log(err);
      return;
    }

    console.log('package.json updated!');
  })
}

function addFile(src, target) {
  return new Promise((resolve, reject) => {
    if (fs.existsSync(target)) {
      resolve('existed');
    }
  
    fs.copyFile(src, target, (err) => {
      if (err) {
        reject(err);
      }
      resolve('done')
    })
  });
}

function addFileContainer(options = {}) {
  const { useTypeScript, isLibrary } = options;

  const addPrettier = new Promise((resolve) => {
    const src = path.resolve(__dirname, '..', 'files/.prettierrc');
    const target = path.resolve(process.cwd(), '.prettierrc');

    resolve(addFile(src, target));
  });

  const addProjectConfig = new Promise((resolve) => {
    const src = path.resolve(__dirname, '..', 'files/project.config.example.js');
    const target = path.resolve(process.cwd(), 'project.config.js');

    resolve(addFile(src, target));
  });

  const addLint = new Promise((resolve) => {
    let filename;
    if (isLibrary) {
      filename = 'files/.eslintrc.lib.json';
    } else if (useTypeScript) {
      filename = 'files/.eslintrc.react-ts.json';
    } else {
      filename = 'files/.eslintrc.react.json';
    }

    const src = path.resolve(__dirname, '..', filename);
    const target = path.resolve(process.cwd(), '.eslintrc.json');

    resolve(addFile(src, target));
  });

  const addTSConfig = new Promise((resolve) => {
    let filename;
    if (isLibrary) {
      filename = 'files/tsconfig.lib.json';
    } else if (useTypeScript) {
      filename = 'files/tsconfig.project.json';
    } else {
      resolve('done');
    }

    const src = path.resolve(__dirname, '..', filename);
    const target = path.resolve(process.cwd(), 'tsconfig.json');

    const srcBaseConfig = path.resolve(__dirname, '..', 'files/tsconfig.base.json');
    const targetBaseConfig = path.resolve(process.cwd(), 'tsconfig.base.json');

    resolve(Promise.all([addFile(src, target), addFile(srcBaseConfig, targetBaseConfig)]));
  });

  const addBabelrc = new Promise((resolve) => {
    if (isLibrary) {
      resolve('done');
    } else {
      const src = path.resolve(__dirname, '..', 'files/.babelrc');
      const target = path.resolve(process.cwd(), '.babelrc');
  
      resolve(addFile(src, target));
    }
  });

  const addCommitlint = new Promise((resolve) => {
    const src = path.resolve(__dirname, '..', 'files/commitlint.config.js');
    const target = path.resolve(process.cwd(), 'commitlint.config.js');
  
    resolve(addFile(src, target));
  });

  Promise.all([addPrettier, addProjectConfig, addLint, addTSConfig, addBabelrc, addCommitlint]).then((values) => {
    console.log('Files added: ', values.toString());
  }).catch((err) => {
    console.log(err);
  })
}

function init(options = {}) {
  addFileContainer(options);
  updatePackage(options);
}

module.exports = init;
